add dplyr::rename

eoduniyi opened the pull request:
Implemented dplyr::rename with unit test, associated documentation, and vingette explanation.

eoduniyi committed:
add dplyr::rename

Implemented dplyr::rename with unit test, associated documentation, and vingette explanation.

eoduniyi committed:
add basic unittest for select.R

Added basic unit test for the select.R function.

cbeleites commented on R/select.R:
```diff
@@ -39,3 +39,48 @@ select.hyperSpec <- function(.data, ...) {
+
+  # UT1
+  test_that("selecting data columns correctly", {
+    df <- data.frame(a = NA, b = NA)
```
UT1 and UT2 test `select()` for data.frame rather than for **hyperSpec** objects. 

Behaviour for data.frame is responsibility of **dplyr** rather than **hyperSpec.tidyverse** and so far we don't have a special reason to check this "foreign" behaviour in **hyperSpec.tidyverse**, so I suggest to drop these 2 tests from select.R.

cbeleites commented on R/select.R:
```diff
@@ -39,3 +39,48 @@ select.hyperSpec <- function(.data, ...) {
+      select(chondro, filename, clusters),
+      data.frame(filename = chondro@data$filename, y = chondro@data$clusters, stringsAsFactors = FALSE)
+    )
+    expect_equivalent(
```
The `tmp` assigned above is really only needed down here. Please move it directly before this `expect_*` to increase readability.

cbeleites commented on R/select.R:
```diff
@@ -39,3 +39,48 @@ select.hyperSpec <- function(.data, ...) {
+  test_that("dropping spectra column", {
+    tmp <- chondro@data
+    tmp$spc <- NULL
+    expect_equivalent(
```
(This is the one really **important** comment in this code review:)

You use `expect_equivalent()` rather than `expect_equal()` - presumably because the test would not pass with `expect_equal()`.   
The issue I see here at the moment is that a single `expect_equivalent()` is too lax to give a good test for `select()`. As an example, consider that the subtest for `select(chondro, filename, clusters)` contains a typo (the 2nd column of the resulting data.frame should be `$clusters`, not `$y`) and in consequence _should_ fail right now. But it doesn't. 

What to do? Here are some alternatives (in the order of preference):

1. add the `"labels"` attribute to the manually constructed result data.frame and use `expect_equal()`.

2. In addition to the `expect_equivalent()` which checks the contents of the resulting data.frame, add two more expectations for crucial behaviour of the result:  an expectation for the column names, and an expectation for the `"labels"` attribute because `select.hyperSpec()` explicitly sets this attribute. 

3. Modify the result of `select.hyperSpec()` so that `expect_equal()` can be used. IMHO, this already needs very careful consideration, i.e. a reason why not to use one of the alternatives above in this particular case.

4. Use `expect_equivalent()` for a weak check (as it is now)

eoduniyi replied:
I have considered your suggestion .3 and it has led me down the following line of thinking:

-> expect_equal checks for equality by comparing values and attributes. 
-> select.hyperSpec ensures that the result has the correct attributes by assigning it the labels of the parameter passed in. 

Which led me to ask the question:

Why is it necessary to assign ALL the labels of the parameter to the result when users are selecting only a subset of the data.frame ```select(chondro, x, y)```? To be clearer, the result generated from ```select(chondro, x, y)``` would still have the labels "filename", "clusters", "spc", and "wavelength", but shouldn't it only have the labels "x" and "y"?

cbeleites replied:
You are right that it should not have `labels$filename`  nor `labels$clusters` - good catch! 

Whether labels `labels$.wavelength` and `labels$spc` should be retained (as the labels belonging to the 0 x 0 matrix `@data$spc` is debatable. I suggest to return labels that are consistent with what `chondro [, c("x", "y")]` does: that retains the `.wavelength` label, but not the `spc` label. I'll push a fix.

cbeleites reviewed (changes requested):
Hi Erick, many thanks for submitting the unit test. I'd like to see a few changes, of which one is important to make this a good unit test.

eoduniyi committed:
update unittest for rename.R and select.R

cbeleites commented on R/rename.R:
```diff
@@ -24,7 +24,6 @@
   res <- dplyr::rename(.data@data, ...)
-  res
   # Not sure if I need to check if 'spc' has been renamed
 }
```
do `chk.hy(res)` here (and do not forget to return `res`)

cbeleites commented on R/rename.R:
```diff
@@ -61,5 +60,10 @@ rename.hyperSpec <- function(.data, ...){
 
+  # UT4
+  test_that("renaming spectrum column throws error", {
+    expect_error(chk.hy(rename(chondro, spc_newcolname = spc)))
```
`chk.hy()` is not needed here - should be done inside `rename.hyperSpec` since otherwise an invalid hyperSpec object may be returned.

cbeleites commented:
Hi @eoduniyi ,

I just pushed a commit to fix the attributes of the data.frame returned by `rename.hyperSpec`. I also added a unit test to make sure the attributes are as supposed.

eoduniyi committed:
Merge branch 'master' into master

eoduniyi committed:
update rename.R

eoduniyi committed:
Revert "update rename.R"

This reverts commit 49761facbcc62873fc4aff9b4322a05842db0aad.

eoduniyi committed:
update rename.R

eoduniyi committed:
update select.R

add newline to end of file.

eoduniyi commented:
Dear @cbeleites as you may have noticed I'm having some trouble with unittesting and understanding the basics of hyperSpec objects. Also, it appears that Travis CI keeps failing on this branch after I make updates to it...I was wondering if you could explain that. I'm not super familiar with Travis TBH. 

Finally, now when I try to ```load_all()``` or build my local copy of hyperSpec.tidyverse I get the following error ```Error: object ‘flu’ is not exported by 'namespace:hyperSpec'```, which I believe is referencing filter.R. However, I checked the filter.R file and the NAMESPACE file and they both have ```importFrom hyperSpec flu``` and ```importFrom(hyperSpec, flu)```?
<img width="1436" alt="Screen Shot 2020-03-27 at 9 30 09 PM" src="https://user-images.githubusercontent.com/31193655/77824509-2cc7f880-70d1-11ea-9e69-546590419cac.png">

eoduniyi committed:
update rename.R and R/

eoduniyi reviewed (commented):
Reviewed changes: looks good!

eoduniyi reviewed (commented):
Reviewed changes: Will implement!

eoduniyi committed:
update rename.R

eoduniyi reviewed (commented):
Reviewed changes: looks good!

eoduniyi commented:
Dear @cbeleites,

I have updated rename.R implementation and its associated unit test. I used `covr` to check how well the test covered rename.R([see img](https://storage.googleapis.com/root-proposal-1246/opensource/RGSoC2020/hyperSpec.test-coverage.png)). 

On one hand, I have a feeling %100 coverage should be taken with a grain of salt. On the other hand, I am having a hard time coming up with more unit tests atm. So as always, if you have any suggestions that would be awesome. 

Also, I took a look at Travis and realized why it kept failing (something related to the examples), fixed them, and so it seems to pass the checks now. Finally, my notes for this test can be viewed [here](https://github.com/eoduniyi/hyperSpec.gsoc2020) under section 3. 

Best,
EO

eoduniyi committed:
add transmute.hyperSpec

Fortified dplyr::transmute (#7), added
unit test, associated documentation, and vingette explanation.

eoduniyi committed:
update DESCRIPTION

update DESCRIPTION to import rlang for transmute.R

eoduniyi committed:
update transmute.R

eoduniyi committed:
update transmute.R

eoduniyi committed:
add hy_update_labels()

added hy_update_labels.R to ensure labels are updated correctly...added to transmute.R and select.R

eoduniyi committed:
Merge branch 'master' into master

cbeleites reviewed (commented):
+1 

One question to all, though: is `hy_update_labels()` a good name?

- Possible confusion with function to actively (by user) set new labels?
- Would `hy_fix_labels()` or `hy_keep_labels()` be more to the point?

eoduniyi commented:
> +1
> 
> One question to all, though: is `hy_update_labels()` a good name?
> 
> * Possible confusion with function to actively (by user) set new labels?
> * Would `hy_fix_labels()` or `hy_keep_labels()` be more to the point?

I think `hy_fix_labels()` is more to the point.
